myokit/_model_api.py

```diff
@@ -1659,6 +1659,276 @@ def has_warnings(self):
+        # print(relevant_vars)
+
+        copied_var_list = []
+
+        # map has form {external_model_copy.variable: self.variable}
+        if var_map is not None:
+            if not isinstance(var_map, dict):
+                raise TypeError('var_map needs to be of type dict or None')
+            for ext_var, self_var in var_map.items():
+                # check self variables in var_map are variables and exists
+                # in self
+                if isinstance(self_var, myokit.Variable):
+                    if not self_var.has_ancestor(self):
+                        raise myokit.WellMappedError(
+                            self_var.name() + ' does not exist in this model'
+                        )
+                elif isinstance(self_var, basestring):
+                    try:
+                        self_var = self.var(self_var)
+                    except KeyError:
+                        raise myokit.WellMappedError(
+                            self_var + ' does not exist in this model'
+                        )
+                else:
+                    raise TypeError(
+                        'Variables in the var_map need to be specified as ' +
+                        'objects or as fully qualified names'
+                    )
+
+                # check external variables in var_map are variables and exists
+                # in external model
+                if isinstance(ext_var, myokit.Variable):
+                    if not ext_var.has_ancestor(external_component.model()):
+                        raise myokit.WellMappedError(
+                            ext_var.name() +
+                            ' does not exist in the external component\'s' +
+                            'model'
+                        )
+                    ext_var = ext_model_copy.var(ext_var.qname())
+                elif isinstance(ext_var, basestring):
+                    try:
+                        ext_var = ext_model_copy.var(ext_var)
+                    except KeyError:
+                        raise myokit.WellMappedError(
+                            ext_var +
+                            ' does not exist in the external component\'s' +
+                            'model'
+                        )
+                else:
+                    raise TypeError(
+                        'Variables in the var_map need to be specified as ' +
+                        'objects or as fully qualified names'
+                    )
+
+                # check for duplicates in var_map
+                if self_var in copied_var_list:
+                    raise myokit.WellMappedError(
+                        'Multiple variables map onto ' + self_var.name()
+                    )
+
+                # add to full_var_map if a relevant variable
+                if ext_var in relevant_vars:
+                    full_var_map[ext_var] = self_var
+                copied_var_list.append(self_var)
+
```

Comment:
Add some comment saying what the next bit does

Reply:
comments have been updated and added throughout, hopefully it's more readable now.